pkg/server/deploy/deploy.go

```diff
@@ -0,0 +1,122 @@
+package deploy
+
+import (
+	"fmt"
+	"io"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/luizalabs/teresa-api/models/storage"
+	"github.com/luizalabs/teresa-api/pkg/server/app"
+	"github.com/luizalabs/teresa-api/pkg/server/auth"
+	st "github.com/luizalabs/teresa-api/pkg/server/storage"
+	"github.com/pborman/uuid"
+)
+
+type Operations interface {
+	Deploy(user *storage.User, appName string, tarBall io.ReadSeeker, description string, rhl int) (io.ReadCloser, error)
+}
+
+type K8sOperations interface {
+	PodRun(podSpec *PodSpec) (io.ReadCloser, <-chan int, error)
+	CreateDeploy(deploySpec *DeploySpec) error
+	HasService(namespace, name string) (bool, error)
+	CreateService(namespace, name string) error
+}
+
+type DeployOperations struct {
+	appOps      app.Operations
+	fileStorage st.Storage
+	k8s         K8sOperations
+}
+
+func (ops *DeployOperations) Deploy(user *storage.User, appName string, tarBall io.ReadSeeker, description string, rhl int) (io.ReadCloser, error) {
+	a, err := ops.appOps.Meta(appName)
+	if err != nil {
+		return nil, err
+	}
+
+	teamName, err := ops.appOps.TeamName(appName)
+	if err != nil {
+		return nil, err
+	}
+	a.Team = teamName
+
+	if !ops.appOps.HasPermission(user, appName) {
+		return nil, auth.ErrPermissionDenied
+	}
+
+	tYaml, err := getTeresaYamlFromTarBall(tarBall) // get tYaml and parse before update deploy
+	if err != nil {
+		return nil, err
+	}
+
+	deployId := genDeployId()
+	buildDest := fmt.Sprintf("deploys/%s/%s/out", appName, deployId)
+
+	r, w := io.Pipe()
+	go func() {
+		defer w.Close()
+		if err = ops.buildApp(tarBall, a, deployId, buildDest, w); err != nil {
+			return
+		}
+		slugURL := fmt.Sprintf("%s/slug.tgz", buildDest)
+		if err := ops.createDeploy(a, tYaml, description, slugURL, rhl); err != nil {
+			log.WithError(err).Errorf("Creating deploy")
+			return
+		}
+
+		ops.exposeService(a, w)
+	}()
+	return r, nil
+}
+
+func (ops *DeployOperations) createDeploy(a *app.App, tYaml *TeresaYaml, description, slugPath string, rhl int) error {
+	deploySpec := newDeploySpec(a, tYaml, ops.fileStorage, description, slugPath, a.ProcessType, rhl)
+	return ops.k8s.CreateDeploy(deploySpec)
+}
+
+func (ops *DeployOperations) exposeService(a *app.App, w io.Writer) {
+	if a.ProcessType != app.ProcessTypeWeb {
+		return
+	}
+	hasSrv, err := ops.k8s.HasService(a.Name, a.Name)
+	if err != nil {
+		log.WithError(err).Errorf("Checking APP service")
+		return
+	}
+	if !hasSrv {
+		fmt.Fprintln(w, "Exposing LoadBalancer service")
+		if err := ops.k8s.CreateService(a.Name, a.Name); err != nil {
```

Comment:
Yes, you're right but today we only deal with `LoadBalancer` service, right?
I not sure if we need to deal with different types of service at this moment, what do you think?

The function `pkg/server/k8s/converters.go/serviceSpec` has an argument called `srvType` and uses it on service specification, We'll need to delegate that information (service type of app) to the app layer though